First lmer try

datalorax opened the pull request:
This is not complete, but if you look at may last comment from #22 you'll see that all equations basically have a "distributed as" part and a variance/covariance part. This PR is the first try at getting the "distributed as" part. Only part way there, but I think it's a pretty good step. Examples below. I put all of this in a new R file for now.

For some reason the variable ordering is not holding. I'll look into that and get that fixed before we merge.

``` r
library(tidyverse)
library(lme4)

data("benchmarks", package = "esvis")

d <- benchmarks %>% 
  add_count(sid) %>% 
  filter(n == 3) %>% 
  mutate(wave = case_when(season == "Fall" ~ 0, 
                          season == "Winter" ~ 1,
                          TRUE ~ 2))

set.seed(123)
d$district <- rep(sample(1:9), length.out = nrow(d))

d <- d %>% 
  count(district, frl) %>% 
  pivot_wider(names_from = frl, values_from = n) %>% 
  mutate(district_prop_frl = FRL / (FRL + `Non-FRL`)) %>% 
  select(district, district_prop_frl) %>% 
  right_join(d) %>% 
  rename(sid_frl = frl) 

m0 <- lmer(math ~ 1 + (1|sid), d)
equatiomatic:::create_distributed_as_merMod(m0, ital_vars = FALSE) %>% 
  cat("$$", ., "$$")
```

![Screen Shot 2020-09-01 at 1 22 04 PM](https://user-images.githubusercontent.com/10944136/91902162-23c9ed00-ec56-11ea-8e85-01ae80ef7c0e.png)

```r
m1 <- lmer(math ~ wave + (1|sid), d)
equatiomatic:::create_distributed_as_merMod(m1, ital_vars = FALSE) %>% 
  cat("$$", ., "$$")
```

![Screen Shot 2020-09-01 at 1 22 34 PM](https://user-images.githubusercontent.com/10944136/91902214-3512f980-ec56-11ea-8eca-1cee2c6bcd2b.png)

```r
m2 <- lmer(math ~ wave + (wave|sid), d)
equatiomatic:::create_distributed_as_merMod(m2, ital_vars = FALSE) %>% 
  cat("$$", ., "$$")
```
![Screen Shot 2020-09-01 at 1 23 15 PM](https://user-images.githubusercontent.com/10944136/91902272-4d831400-ec56-11ea-8f6f-521d3d5fc27c.png)

```r
m3 <- lmer(math ~ wave + cohort + (wave|sid), d)
equatiomatic:::create_distributed_as_merMod(m3, ital_vars = FALSE) %>% 
  cat("$$", ., "$$")
```

![Screen Shot 2020-09-01 at 1 28 37 PM](https://user-images.githubusercontent.com/10944136/91902785-16f9c900-ec57-11ea-86d4-86df7176ffd4.png)

```r
m4 <- lmer(math ~ wave + cohort + (wave|sid) + (wave + cohort|district), d)
#> boundary (singular) fit: see ?isSingular
equatiomatic:::create_distributed_as_merMod(m4, ital_vars = FALSE) %>% 
  cat("$$", ., "$$")
```

![Screen Shot 2020-09-01 at 1 25 21 PM](https://user-images.githubusercontent.com/10944136/91902447-9935bd80-ec56-11ea-84e2-e8073f6a00a3.png)

```r
m5 <- lmer(math ~ wave*cohort + (wave|sid) + (wave*cohort|district), d)
#> boundary (singular) fit: see ?isSingular
equatiomatic:::create_distributed_as_merMod(m5, ital_vars = FALSE) %>% 
  cat("$$", ., "$$")
```

![Screen Shot 2020-09-01 at 1 26 06 PM](https://user-images.githubusercontent.com/10944136/91902509-b36f9b80-ec56-11ea-98e7-e73e77db07f2.png)

datalorax commented:
Okay, the misordering was coming from a merge. It should be fixed now.

datalorax commented:
ah shoot... also realized there's some code formatting that needs to happen here. I went well past the 80 character line. Will fix before we merge. Trying to work on the var-covar part now.

datalorax commented:
I'm done for the day but just wanted to share this quickly (haven't pushed any of this yet). 

```r
m5 <- lmer(math ~ wave * sid_frl + cohort + district_prop_frl + (wave|sid) + (wave + cohort + wave:sid_frl|district), d)
```

Overly complicated model that doesn't fit, but, with my local version anyway, you can now do this!

```r
equatiomatic:::create_vcov_structure_merMod(model) %>% 
  cat("$$", ., "$$")
```

Which renders as

![Screen Shot 2020-09-01 at 9 27 21 PM](https://user-images.githubusercontent.com/10944136/91932174-43830480-ec9a-11ea-8fe9-be1891fdf63a.png)


The actual LaTeX for this is kind of crazy. It ends up looking like this.

```
\left(
       \begin{array}{c} \alpha_{j} \\ \beta_{1j} \\ \beta_{2j} \\ \beta_{3j} \end{array}
     \right)\sim N \left(\left(
       \begin{array}{c} \mu_{\alpha} \\ \mu_{\beta_{1}} \\ \mu_{\beta_{2}} \\ \mu_{\beta_{3}} \end{array}
     \right),\left(
    \begin{array}{cccc}\sigma^2_{\alpha} & \rho\sigma_{\alpha}\sigma_{\beta_{1}} & \rho\sigma_{\alpha}\sigma_{\beta_{2}} & \rho\sigma_{\alpha}\sigma_{\beta_{3}}\\\rho\sigma_{\alpha}\sigma_{\beta_{1}} & \sigma^2_{\beta_{1}} & \rho\sigma_{\beta_{1}}\sigma_{\beta_{2}} & \rho\sigma_{\beta_{1}}\sigma_{\beta_{3}}\\\rho\sigma_{\alpha}\sigma_{\beta_{2}} & \rho\sigma_{\beta_{1}}\sigma_{\beta_{2}} & \sigma^2_{\beta_{2}} & \rho\sigma_{\beta_{2}}\sigma_{\beta_{3}}\\\rho\sigma_{\alpha}\sigma_{\beta_{3}} & \rho\sigma_{\beta_{1}}\sigma_{\beta_{3}} & \rho\sigma_{\beta_{2}}\sigma_{\beta_{3}} & \sigma^2_{\beta_{3}}\end{array}
    \right) \right) \\ \left(
       \begin{array}{c} \alpha_{k} \\ \beta_{1k} \end{array}
     \right)\sim N \left(\left(
       \begin{array}{c} \mu_{\alpha} \\ \mu_{\beta_{1}} \end{array}
     \right),\left(
    \begin{array}{cc}\sigma^2_{\alpha} & \rho\sigma_{\alpha}\sigma_{\beta_{1}}\\\rho\sigma_{\alpha}\sigma_{\beta_{1}} & \sigma^2_{\beta_{1}}\end{array}
    \right) \right)
```

But it all builds automatically! Still a fair amount to do, including how to work in group-level predictors (they all go to top level at this point) and building in things like align environments. I'd also like to add the "for SID j in 1,...,J" part for each level. But this is promising!

andrewheiss commented:
zomg this is incredible!

datalorax commented:
Wooohooo!!!! This is all working now. There's still a fair amount of things I'd like to implement but we're basically there for these basic models!

Last thing that needs to be changed before we merge, though, is the codebase needs to be cleaned up. This is really unorganized right now. I'll work on that later today, and then we can merge it in if it all looks good to you? 

``` r
library(equatiomatic)
library(lme4)
#> Loading required package: Matrix
library(tidyverse)

data("benchmarks", package = "esvis")

d <- benchmarks %>% 
  add_count(sid) %>% 
  filter(n == 3) %>% 
  mutate(wave = case_when(season == "Fall" ~ 0, 
                          season == "Winter" ~ 1,
                          TRUE ~ 2))

set.seed(123)
d$district <- rep(sample(1:9), length.out = nrow(d))

d <- d %>% 
  count(district, frl) %>% 
  pivot_wider(names_from = frl, values_from = n) %>% 
  mutate(district_prop_frl = FRL / (FRL + `Non-FRL`)) %>% 
  select(district, district_prop_frl) %>% 
  right_join(d) %>% 
  rename(sid_frl = frl) 
#> Joining, by = "district"

m0 <- lmer(math ~ 1 + (1|sid), d)
extract_eq(m0)
#> Registered S3 method overwritten by 'broom.mixed':
#>   method      from 
#>   tidy.gamlss broom
```

![Screen Shot 2020-09-02 at 12 32 58 PM](https://user-images.githubusercontent.com/10944136/92028158-7075fc80-ed18-11ea-84f1-f96d5b6d2c4d.png)

``` r

m1 <- lmer(math ~ 1 + wave + (1|sid), d)
extract_eq(m1)
```

![Screen Shot 2020-09-02 at 12 33 55 PM](https://user-images.githubusercontent.com/10944136/92028242-926f7f00-ed18-11ea-982d-b8bf30e6a856.png)


``` r
m2 <- lmer(math ~ 1 + wave + (wave|sid), d)
extract_eq(m2)
```

![Screen Shot 2020-09-02 at 12 33 19 PM](https://user-images.githubusercontent.com/10944136/92028190-7d92eb80-ed18-11ea-9360-80bdd38c1d6b.png)

``` r
m3 <- lmer(math ~ 1 + wave + (wave|sid) + (1|district), d)
#> boundary (singular) fit: see ?isSingular
extract_eq(m3)
```

![Screen Shot 2020-09-02 at 12 34 16 PM](https://user-images.githubusercontent.com/10944136/92028275-9ef3d780-ed18-11ea-964b-54ef7c4897f3.png)


``` r
m4 <- lmer(math ~ 1 + wave*cohort + (wave|sid) + (wave*cohort|district), d)
#> boundary (singular) fit: see ?isSingular
extract_eq(m4)
```
![Screen Shot 2020-09-02 at 12 34 44 PM](https://user-images.githubusercontent.com/10944136/92028314-b03ce400-ed18-11ea-9922-09a691d6988d.png)

jrosen48 commented:
Magical. Kicking the tires, on this branch:

```{sh}
On branch first-lmer-try
Your branch is up to date with 'origin/first-lmer-try'.
```

After building the package, I got what is probably a super easy to fix namespace error for `stringr::str_split()`:

``` r
library(lme4)
#> Loading required package: Matrix
library(equatiomatic)

fm1 <- lmer(Reaction ~ Days + (Days | Subject), sleepstudy)

extract_eq(fm1)
#> Registered S3 method overwritten by 'broom.mixed':
#>   method      from 
#>   tidy.gamlss broom
#> Error in str_split(text, " "): could not find function "str_split"
```

<sup>Created on 2020-09-02 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

datalorax commented:
Ah shoot! I meant to use `strsplit` instead of `str_split`. Just pushed a fix.

jrosen48 commented:
``` r
library(lme4)
#> Loading required package: Matrix
library(equatiomatic)

fm1 <- lmer(Reaction ~ Days + (Days | Subject), sleepstudy)

extract_eq(fm1)
#> Registered S3 method overwritten by 'broom.mixed':
#>   method      from 
#>   tidy.gamlss broom
```

\[
\begin{aligned}
\operatorname{Reaction}  &\sim N \left(\alpha_{j[i]} + \beta_{1j[i]}(\operatorname{Days}),\sigma^2 \right) \\ \left(
       \begin{array}{c} \alpha_{j} \\ \beta_{1j} \end{array}
     \right) &\sim N \left(\left(
       \begin{array}{c} \mu_{\alpha} \\ \mu_{\beta_{1}} \end{array}
     \right),\left(
    \begin{array}{cc}\sigma^2_{\alpha} & \rho \sigma_{\alpha} \sigma_{\beta_{1}}\\\rho \sigma_{\beta_{1}} \sigma_{\alpha} & \sigma^2_{\beta_{1}}\end{array}
    \right) \right) , \text{ for  Subject } j  = 1, \dots ,  J
\end{aligned}
\]

<sup>Created on 2020-09-02 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

Seriously magical. I will try a few more examples. My challenge now is not fully understanding the Gelman and Hill notation... wondering if some kind of documentation around how to "translate" between the Raudenbush & Bryk-esque notation and it. I guess on that note, perhaps as a newcomer to this notation I can find the section of *Data Analysis Using Regression and Multilevel/Hierarchical Models* where they show five (or six?) ways to write the same model...

datalorax commented:
Thanks, that would be super helpful. Also - I realize the tests are failing. I'll fix that later tonight too when I organize the codebase a bit better.

andrewheiss commented:
This is absolutely amazing.

datalorax commented:
Okay, I just documented all the functions and cleaned up the codebase some (gave better names to a few functions, etc.). The next thing I really want to implement is being able to define group-level coefficients. I have a few ideas on that, but I think this probably ready to merge in?

datalorax commented:
Okay, I refactored the code and fixed a couple small bugs. I also added wrapping and the ability to have the mean structure inside or outside of the normal distribution part. I called that argument `mean_separate`, which I don't love and am fully open to suggestions for other names.

I will wait until this evening but I do believe this is officially ready to merge, and then we can submit new PRs for further development (e.g. #105). 

```r
library(equatiomatic)
library(lme4)
library(tidyverse)
data("benchmarks", package = "esvis")

d <- benchmarks %>% 
  add_count(sid) %>% 
  filter(n == 3) %>% 
  mutate(wave = case_when(season == "Fall" ~ 0, 
                          season == "Winter" ~ 1,
                          TRUE ~ 2))

set.seed(123)
d$district <- rep(sample(1:9), length.out = nrow(d))

d <- d %>% 
  count(district, frl) %>% 
  pivot_wider(names_from = frl, values_from = n) %>% 
  mutate(district_prop_frl = FRL / (FRL + `Non-FRL`)) %>% 
  select(district, district_prop_frl) %>% 
  right_join(d) %>% 
  rename(sid_frl = frl) 

m0 <- lmer(math ~ 1 + (1|sid), d)
extract_eq(m0)
```

![Screen Shot 2020-09-04 at 2 57 05 PM](https://user-images.githubusercontent.com/10944136/92287541-e7e59080-eebe-11ea-9f29-624e6178db69.png)


```r
m1 <- lmer(math ~ 1 + wave + (1|sid), d)
extract_eq(m1)
```

![Screen Shot 2020-09-04 at 2 57 21 PM](https://user-images.githubusercontent.com/10944136/92287557-f3d15280-eebe-11ea-94a8-dcc0c62593b3.png)

```r
m2 <- lmer(math ~ 1 + wave + (wave|sid), d)
extract_eq(m2)
```

![Screen Shot 2020-09-04 at 2 58 03 PM](https://user-images.githubusercontent.com/10944136/92287587-0a77a980-eebf-11ea-90e6-c404a1ccc125.png)

```r
m3 <- lmer(math ~ 1 + wave + (wave|sid) + (1|district), d)
extract_eq(m3)
```

![Screen Shot 2020-09-04 at 2 58 33 PM](https://user-images.githubusercontent.com/10944136/92287614-1c594c80-eebf-11ea-8942-dc51208100f2.png)

```r
m4 <- lmer(math ~ 1 + wave*cohort + (wave|sid) + (wave*cohort|district), d)
extract_eq(m4)
```

![Screen Shot 2020-09-04 at 2 59 04 PM](https://user-images.githubusercontent.com/10944136/92287636-2ed38600-eebf-11ea-93ce-08c40aff252d.png)

## Show wrapping

```r
m5 <- lmer(math ~ wave*cohort + district_prop_frl + 
            (wave|sid) + (wave*cohort|district),
           data = d)  
extract_eq(m5, wrap = TRUE, terms_per_line = 2)
```

![Screen Shot 2020-09-04 at 2 59 37 PM](https://user-images.githubusercontent.com/10944136/92287666-43178300-eebf-11ea-9b33-22abb04a5260.png)

### Show wrapping inside normal distribution
```r
extract_eq(m5, wrap = TRUE, terms_per_line = 2, mean_separate = FALSE)
```

![Screen Shot 2020-09-04 at 3 00 00 PM](https://user-images.githubusercontent.com/10944136/92287687-4f9bdb80-eebf-11ea-8147-72059a521961.png)

bbolker commented:
Wow! I'll take a look.

If you need to access `tidy.merMod` directly, `broom.mixed` could certainly export it, although it's not immediately obvious to me why regular S3 dispatch doesn't do what you need ... ?

bbolker commented:
> It's not a huge deal, but because already use `broom::tidy` I get the following message when running `extract_eq()` with a `merMod`. Maybe there's another way around this though?
> 
> > Registered S3 method overwritten by 'broom.mixed':
> > method      from
> > tidy.gamlss broom

The correct way to get around this would be to have the `broom` and `broom.mixed` maintainers agree that only one of them was going to export such a method.  Based on [this forgotten issue](https://github.com/bbolker/broom.mixed/issues/64), it looks like I should remove `tidy.gamlss` from `broom.mixed` ...

datalorax commented:
Went ahead and merged this in. I have a nearly working branch on my local to incorporate group-level predictors. Will have a PR for that soon.
